src/code_wars/sg1.py

```diff
@@ -0,0 +1,112 @@
+import math
+from queue import PriorityQueue
+
+
+class DialHomeDevice:
+    def __init__(self, galactic_map):
+        self.start, self.goal = None, None
+        self._galactic_gates = [
+            [self._create_gate(x, y, gate) for x, gate in enumerate(gates)]
+            for y, gates in enumerate(galactic_map.split("\n"))
+        ]
+        self._set_connections()
+
+    def _create_gate(self, x, y, gate_key):
+        if gate_key == "X":
+            return None
+        gate = Gate(x, y)
+        if gate_key == "S":
+            self.start = gate
+        elif gate_key == "G":
+            self.goal = gate
+        return gate
+
+    def _set_connections(self):
+        for gates in self._galactic_gates:
+            for gate in gates:
+                if gate:
+                    gate.connect(self._galactic_gates, self.goal)
+
+    def dial_home(self):
+        queue = PriorityQueue()
+        queue.put((0, self.start, []))
+        visited = {}
+        visited[self.start] = 0
+        shortest_route = None
+        shortest_distance = math.inf
+        while queue.qsize() > 0:
+            previous_distance, current, route = queue.get_nowait()
+            for gate in current.connections:
+                distance = previous_distance + math.dist(
+                    gate.location, current.location
+                )
+                if gate == self.goal:
+                    if distance < shortest_distance:
+                        shortest_route = route
+                        shortest_distance = distance
+                else:
+                    visited_distance = visited.get(gate, math.inf)
+                    if visited_distance > distance:
+                        queue.put((distance, gate, route + [gate]))
+                        visited[gate] = distance
+        if shortest_distance < math.inf:
+            return self._route(shortest_route)
+        return "Oh for crying out loud..."
+
+    def _route(self, route):
+        map_key = {None: "X", self.start: "S", self.goal: "G"}
+        return "\n".join(
+            [
+                "".join(
+                    [
+                        "P" if gate and gate in route else map_key.get(gate, ".")
+                        for gate in gates
+                    ]
+                )
+                for gates in self._galactic_gates
+            ]
+        )
+
+
+class Gate:
+    def __init__(self, x, y):
+        self.location = x, y
+        self.connections = set()
+        self.distance_from_goal = None
+
+    def __hash__(self):
+        return hash(self.location)
+
+    def __eq__(self, other_gate):
+        return self.location == other_gate.location
+
+    def __lt__(self, other_gate):
+        return self.distance_from_goal < other_gate.distance_from_goal
```

Comment:
Why do I need these?

Reply:
I am still not there yet with an answer on why to use __hash__ at all but I feel like I am close.

But I do know that since you are using __hash__, you then have to use __eq__ otherwise you'd risk the hash not being a true hash? 

So, I assume you are wanting to use a hash so you can make your list as fast as a set?

But having said that, I still am going over and over it to understand what is happening so don't accidentally give too much away with your response. Unless I am way off, then drop a subtle hint.

Reply:
is the __lt__ used to evaluate what goes first in PriorityQueue()? So .distance_from_goal that is less gets higher priority?

Reply:
forgive me for rambling instead of waiting till I understand it more but I just gotta ramble...

is the `__hash__` being used because two gates can share the same location? So, the hash lets you identify them uniquely even if their location is equal?

Reply:
All very good educated guesses and close!  I wanted my Gates to be hashable so that I can use them in sets and as keys.  In `visited` and `connections` for example.  I think that they technically would have been hashable without this, but by providing the hashing function, I'm telling Python what it means for a gate to be unique.

In other words because I say the location determines unique ness,
```
gate1 = Gate(0, 0)
gate2 = Gate(0, 0)
gate1 == gater2
```
and
```
s = set()
gate1 = Gate(0, 0)
s.add(gate1)
gate2 = Gate(0, 0)
gate2 in s == True
```
If you define hash, you also have to define equal to account for hashing collisions.

--------
The `__lt__` is needed for the priority queue to work.  If 2 priorities are the same, Python checks to see which item in the queue is larger. My code would have worked fine regardless of what I chose to return here but thought all things being equal I might as well return the Gate closest to the goal as a crow flies.